README.md
# Joomla! Vimeo Custom Field Package

![GitHub release](https://img.shields.io/github/v/release/piotr-cz/joomla-customfields-vimeo)
![Joomla 3](https://img.shields.io/badge/joomla-3.7+-007ec6.svg)
![Joomla 4](https://img.shields.io/badge/joomla-4.0--beta.7+-007ec6.svg)
![PHP](https://img.shields.io/badge/php-7.1+-8892BF.svg)

This plugins lets you create new fields of type 'Vimeo' in any extensions where custom fields are supported.

Package contains following extensions:

- **[Vimeo Fields plugin](#vimeo-fields-plugin)**

  This plugin adds ability to add Vimeo to an article (or any other supported content) by defining Vimeo URL.

- **[Vimeo Progress Content plugin](#vimeo-progress-content-plugin)**

  This plugin adds ability to indicate seen Vimeo videos for logged-in users.  
  It does it by adding
  
  - symbol indicator after an article title (✓)

  - progress bar after category and subcategory title (<progress max="5" value="2" title="2/ 5">40%</progress>)


## Requirements

- Joomla 3.7+ or 4.0-beta.7+

- PHP 7.1+


## Installation

1. Install

   - by pasting URL to zip file of [latest release](https://github.com/piotr-cz/joomla-customfields-vimeo/releases/latest)  
     *Extensions > Manage > Install > Install from URL*

   - by downloading [latest release](https://github.com/piotr-cz/joomla-customfields-vimeo/releases/latest) and installing it using Extension Manager  
     *Extensions > Manage > Install > Upload Package File*

1. Enable Plugins  
   *Extensions > Plugins > Fields - Pcz - Vimeo > Enable*  
   *Extensions > Plugins > Content - Pcz - Vimeo Progress > Enable* (Optional)


## Vimeo Fields plugin

### Configuration

Custom field may be configured either

- globally in plugin  
  *Extensions > Plugins > Fields - Pcz - Vimeo*

- per each field


#### Available settings

- Aspect ratio

- Disable on category views

- Finished videos data store

- All Vimeo Player Parameters (See *Vimeo Help Center > [Using Player Parameters](https://vimeo.zendesk.com/hc/en-us/articles/360001494447-Using-Player-Parameter)*)

  Notes:

  - Some parameters are available only on Vimeo Plus account or higher (*Background*, *Color*, *Controls*, *Quality*, *Speed*).

  - Some parameters have effect only if video owner allows (*Byline*, *Portrait*, *Title*).

    See *Vimeo admin panel > Videos > [video] > Advanced > Embed > Your details > Let users decide*


## Vimeo progress Content plugin

### Configuration

First of all, you have create storage where data about videos seen by users will be saved to.
Plugin doesn't create it's own storage (like database table) but uses custom fields within user context.

To add finished Videos data store, add new field with type *Vimeo Datastore*

1. *Users > Fields > New*

1. Title: *Videos seen by users* (or any custom name)

1. Type: *Vimeo Datastore*

1. Optional: If you want to hide data from users, you may restrict *Access* level, set *Options > Editable in: Site* or set *Options > Display when Read-Only: No*

Now set up Vimeo Fields plugin to use the storage you just created:

1. *Extensions > Plugins > Fields - Pcz - Vimeo*

1. Finished videos data store: *Videos seen by users* (or custom name you entered)


### Available settings

- Enable for category

- Enable for subcategories

- Enable for articles

  - Article seen idicator

  - Article unseen indicator

- Categories to process

- Fields to process


## Testing pre-releases

To allow updates for latest commit switch  
*Global Configuration > Installer > Minimum Extension Stability* to *Development*.

Available extension update will show up as version *99.0.0-alpha* and install as of latest commit from main branch of this repository.


## Known issues

### Joomla 4

- Radio controls in plugin settings look different to Joomla 4 native ones.

  This is a aide effect of having extension compatible with both Jooma 3 and Joomla 4.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.5] - 2022-02-18
### Fixed
- Fix datastore items view when for in case of one value
- J4: Fix assets loading
- J3: Fix reference to asset

## [1.0.4] - 2021-09-28
### Fixed
- Using wrong methods in layout

## [1.0.3] - 2021-09-28 [YANKED]
### Fixed
- Vimeo hash on embeded player

## [1.0.2] - 2021-06-28
### Fixed
- Overlooked package version bump

## [1.0.1] - 2021-06-24
### Fixed
- Fix data store layout (Fields plugin)

### Changed
- Use sha1 for media versions

## [1.0.0] - 2021-06-18
### Added
- Initial release

[Unreleased]: https://github.com/piotr-cz/joomla-customfields-vimeo/compare/v1.0.5...HEAD
[1.0.5]: https://github.com/piotr-cz/joomla-customfields-vimeo/compare/v1.0.4...v1.0.5
[1.0.4]: https://github.com/piotr-cz/joomla-customfields-vimeo/compare/v1.0.3...v1.0.4
[1.0.3]: https://github.com/piotr-cz/joomla-customfields-vimeo/compare/v1.0.2...v1.0.3
[1.0.2]: https://github.com/piotr-cz/joomla-customfields-vimeo/compare/v1.0.1...v1.0.2
[1.0.1]: https://github.com/piotr-cz/joomla-customfields-vimeo/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/piotr-cz/joomla-customfields-vimeo/releases/tag/v1.0.0


pcz_vimeo.php
<?php
/**
 * @package     Joomla.Plugin
 * @subpackage  Fields.pcz_vimeo
 *
 * @copyright   Copyright (C) 2021 Piotr Konieczny. All rights reserved.
 * @license     GNU General Public License version 3 or later; see http://www.gnu.org/licenses/gpl-3.0.txt
 */

/**
 * Joomla 4.x
 * use Joomla\CMS\Form\Form;
 * use Joomla\Component\Fields\Administrator\Plugin\FieldsPlugin;
 */

defined('_JEXEC') or die;

// Joomla 3.x & 4.0
JLoader::import('components.com_fields.libraries.fieldsplugin', JPATH_ADMINISTRATOR);

/**
 * Fields Vimeo Plugin
 * Available Site Events:
 * - onCustomFieldsBeforePrepareField
 * - onCustomFieldsPrepareField
 * - onCustomFieldsAfterPrepareField
 * - onAfterDispatch
 * - onBeforeCompilehead
 * - onAfterRender
 *
 * @since  1.0.0-alpha.1
 */
class PlgFieldsPcz_Vimeo extends FieldsPlugin
{
	/**
	 * @var  boolean
	 *
	 * @inheritdoc
	 */
	protected $autoloadLanguage = true;

	/**
	 * @var \Joomla\CMS\Application\CMSApplication
	 */
	protected $app = null;

	/**
	 * @inheritdoc
	 *
	 * @param   stdClass    $field   The field.
	 * @param   DOMElement  $parent  The field node parent.
	 * @param   JForm       $form    The form.
	 * @return  DOMElement|null
	 */
	public function onCustomFieldsPrepareDom($field, DOMElement $parent, JForm $form): ?DOMElement
	{
		/** @var \DOMElement */
		$fieldNode = parent::onCustomFieldsPrepareDom($field, $parent, $form);

		if (!$fieldNode)
		{
			return $fieldNode;
		}

		// Require helper for filter functions called by JForm validate.
		JLoader::register('JFormRuleVimeoVideoLink', __DIR__ . '/rules/vimeovideolink.php');

		/*
		 * Set field type, filter and validation rule
		 * @see https://docs.joomla.org/Special:MyLanguage/J3.x:Adding_custom_fields/Parameters_for_all_Custom_Fields
		 * @see administrator/components/com_fields/src/Plugin/FieldsPlugin.php
		 */
		$fieldNode->setAttribute('type', 'url');
		$fieldNode->setAttribute('filter', 'JFormRuleVimeoVideoLink::filterVimeoVideoLink');
		$fieldNode->setAttribute('validate', 'VimeoVideoLink');

		// Invalid field: Invalid Video link format
		// $fieldNode->setAttribute('message', 'PLG_FIELDS_PCZ_VIMEO_VALUE_VIDEO_LINK_FIELD_INVALID');

		// Set field options if not filled in admin

		// Label
		if (!$fieldNode->getAttribute('label'))
		{
			$fieldNode->setAttribute('label', 'PLG_FIELDS_PCZ_VIMEO_VALUE_VIDEO_LINK_LABEL');
		}

		// Description
		if (!$fieldNode->getAttribute('description'))
		{
			$fieldNode->setAttribute('description', 'PLG_FIELDS_PCZ_VIMEO_VALUE_VIDEO_LINK_DESC');
		}

		// Placeholder
		if (!$fieldNode->getAttribute('hint'))
		{
			$fieldNode->setAttribute('hint', 'https://vimeo.com/524933864');
		}

		return $fieldNode;
	}

	/**
	 * @inheritdoc
	 *
	 * @param   string    $context  The context.
	 * @param   stdClass  $item     The item.
	 * @param   stdClass  $field    The field.
	 * @return  string|void
	 */
	public function onCustomFieldsPrepareField($context, $item, $field): ?string
	{
		// Check if the field should be processed by us
		if (!$this->isTypeSupported($field->type))
		{
			return null;
		}

		// Merge the params from the plugin and field which has precedence
		$fieldParams = clone $this->params;
		$fieldParams->merge($field->fieldparams);

		$componentView = $this->app->input->get('view');

		// Skip on blog/ list layouts
		if ($fieldParams->get('disable_on_category', 1) && $componentView === 'category')
		{
			return null;
		}

		return parent::onCustomFieldsPrepareField($context, $item, $field);
	}
}


rules/vimeovideolink.php
<?php
/**
 * @package     Joomla.Plugin
 * @subpackage  Fields.pcz_vimeo
 *
 * @copyright   Copyright (C) 2021 Piotr Konieczny. All rights reserved.
 * @license     GNU General Public License version 3 or later; see http://www.gnu.org/licenses/gpl-3.0.txt
 */

use Joomla\Registry\Registry;
use Joomla\Uri\Uri;
use Joomla\CMS\Form\Form;
use Joomla\CMS\Form\Rule\UrlRule;

defined('_JEXEC') or die;

// Require helper
JLoader::register('PlgFieldsPcz_VimeoHelper', __DIR__ . '/../helper.php');

/**
 * JFormVimeolink for plg_fields_pcz_vimeo to make sure value is valid Vimeo video link
 *
 * @since  1.0.0-alpha.1
 */
class JFormRuleVimeoVideoLink extends UrlRule
{
	/**
	 * @inheritdoc
	 *
	 * @param   \SimpleXMLElement  $element  The SimpleXMLElement object representing the `<field>` tag for the form field object.
	 * @param   mixed              $value    The form field value to validate.
	 * @param   string             $group    The field name group control value. This acts as an array container for the field.
	 * @param   Registry           $input    An optional Registry object with the entire data set to validate against the entire form.
	 * @param   Form               $form     The form object for which the field is being tested.
	 *
	 * @return  boolean  True if the value is valid, false otherwise.
	 */
	public function test(\SimpleXMLElement $element, $value, $group = null, Registry $input = null, Form $form = null)
	{
		// Parent tests
		if (!parent::test($element, $value, $group, $input, $form))
		{
			return false;
		}

		// If the field is empty and not required, the field is valid.
		$required = ((string) $element['required'] === 'true' || (string) $element['required'] === 'required');

		if (!$required && empty($value))
		{
			return true;
		}

		$uri = new Uri($value);

		// Scheme
		if ($uri->getScheme() !== 'https')
		{
			return false;
		}

		// Host
		if ($uri->getHost() !== 'vimeo.com')
		{
			return false;
		}

		// Port
		if ($uri->getPort())
		{
			return false;
		}

		// User
		if ($uri->getUser())
		{
			return false;
		}

		// Password
		if ($uri->getPass())
		{
			return false;
		}

		// Path
		if (!$uri->getPath())
		{
			return false;
		}

		// Path params
		$pathParams = PlgFieldsPcz_VimeoHelper::getVimeoPathParams($uri->getPath());

		// Failed to parse format
		if (!is_int($pathParams['vimeoId']))
		{
			return false;
		}

		// Query
		if ($uri->getQuery())
		{
			return false;
		}

		// Fragment
		if ($uri->getFragment())
		{
			return false;
		}

		return true;
	}

	/**
	 * Custom filter/ sanitizer to normalize any Vimeo url to Regular vimeo video link
	 *
	 * @example
	 * ```xml
	 * <field filter="JFormRuleVimeovideolink::filterVimeoVideoLink" />
	 * ```
	 *
	 * Class must be loaded manually bucausse Joomla Form doesn't to so
	 * @example
	 * ```php
	 * // Require helper for filter functions called by JForm.
	 * JLoader::register('JFormRuleVimeovideolink', __DIR__ . '/rules/vimeovideolink.php');
	 * ```
	 *
	 * @param   string  $value  The string to filter
	 * @return  string|null  The filtered string which will be set as a value or null on failure
	 *
	 * @see https://developer.vimeo.com/api/oembed/videos#table-1
	 */
	public static function filterVimeoVideoLink(string $value): ?string
	{
		if (empty($value))
		{
			return $value;
		}

		$uri = new Uri($value);

		// Extract Vimeo ID + Unlisted hash
		$pathParams = PlgFieldsPcz_VimeoHelper::getVimeoPathParams($uri->getPath());

		// Failed to parse format
		if (!is_int($pathParams['vimeoId']))
		{
			return null;
		}

		// Create path with removing optional unlisted hash when null
		$normalizedUrlPath = sprintf('/%s', implode('/', array_filter($pathParams)));

		// Replace path
		$uri->setPath($normalizedUrlPath);

		// Drop query and fragment
		$uri->setQuery(null);
		$uri->setFragment(null);

		return (string) $uri;
	}
}


helper.php
<?php
/**
 * @package     Joomla.Plugin
 * @subpackage  Fields.pcz_vimeo
 *
 * @copyright   Copyright (C) 2021 Piotr Konieczny. All rights reserved.
 * @license     GNU General Public License version 3 or later; see http://www.gnu.org/licenses/gpl-3.0.txt
 */

use Joomla\Registry\Registry;

/**
 * Helper for plg_fields_pcz_vimeo
 *
 * @since  1.0.0-alpha.1
 */
class PlgFieldsPcz_VimeoHelper
{
	/**
	 * Get Vimeo ID from URL
	 *
	 * @param   string  $fieldValue  Vimeo ID or Vimeo URL, ie 'https://vimeo.com/286898202'
	 * @return  integer
	 */
	public static function getVimeoId(string $fieldValue): ?int
	{
		if ($fieldValue == '')
		{
			return null;
		}

		// Legacy - allow Vimeo ID
		if (is_numeric($fieldValue))
		{
			return $fieldValue;
		}

		$urlPath = parse_url($fieldValue, PHP_URL_PATH);
		$urlPathParams = static::getVimeoPathParams($urlPath);

		return $urlPathParams['vimeoId'];
	}

	/**
	 * Get Vimeo parameters that may be used as video src query params
	 *
	 * @param   \Joomla\Registry\Registry  $fieldParams  Field parameters
	 * @return  array
	 *
	 * @see [Vimeo: Using Player Parameters]{@link https://vimeo.zendesk.com/hc/en-us/articles/360001494447-Using-Player-Parameters
	 */
	public static function getVimeoParams(Registry $fieldParams): array
	{
		// Default Vimeo player parameters
		$defaultVimeoParams = [
			'autopause'   => 1,
			'autoplay'    => 0,
			'background'  => 0,
			'byline'      => 1,
			'color'       => '00adef',
			'controls'    => 1,
			'dnt'         => 0,
			'loop'        => 0,
			'muted'       => 0,
			'pip'         => 0,
			'playsinline' => 1,
			'portrait'    => 1,
			'quality'     => 'auto',
			'speed'       => 0,
			'#t'          => '0m',
			'texttrack'   => 0,
			'title'       => 1,
			'transparent' => 1,
		];

		// User defined Vimeo player parameters
		$vimeoParams = [
			'autopause'   => (int) $fieldParams->get('vp_autopause', 1),
			'autoplay'    => (int) $fieldParams->get('vp_autoplay', 0),
			'background'  => (int) $fieldParams->get('vp_background', 0),
			'byline'      => (int) $fieldParams->get('vp_byline', 1),
			'color'       => ltrim((string) $fieldParams->get('vp_color', '#00adef'), '#'),
			'controls'    => (int) $fieldParams->get('vp_controls', 1),
			'dnt'         => (int) $fieldParams->get('vp_dnt', 0),
			'loop'        => (int) $fieldParams->get('vp_loop', 0),
			'muted'       => (int) $fieldParams->get('vp_muted', 0),
			'pip'         => (int) $fieldParams->get('vp_pip', 0),
			'playsinline' => (int) $fieldParams->get('vp_playsinline', 1),
			'portrait'    => (int) $fieldParams->get('vp_portrait', 1),
			'quality'     => (string) $fieldParams->get('vp_quality', 'auto'),
			'speed'       => (int) $fieldParams->get('vp_speed', 0),
			'#t'          => (string) $fieldParams->get('vp_t', '0m'),
			'texttrack'   => (int) $fieldParams->get('vp_texttrack', 0),
			'title'       => (int) $fieldParams->get('vp_title', 1),
			'transparent' => (int) $fieldParams->get('vp_transparent', 1),
		];

		// Remove query params when same as defaults
		return array_filter(
			$vimeoParams,
			function ($value, string $key) use ($defaultVimeoParams) {
				return !array_key_exists($key, $defaultVimeoParams) || $defaultVimeoParams[$key] !== $value;
			},
			ARRAY_FILTER_USE_BOTH
		);
	}

	/**
	 * Get Vimeo URL path parameters by detecting Vimeo URL scheme
	 * @see https://developer.vimeo.com/api/oembed/videos#table-1
	 *
	 * @param   string  $urlPath  URL path component
	 * @return  array
	 */
	public static function getVimeoPathParams(string $urlPath): array
	{
		list ($firstSegment) = explode('/', ltrim($urlPath, '/'), 2);

		switch ($firstSegment)
		{
			// Showcase
			case 'album':
				list ($albumId, $vimeoId, $unlistedHash) = sscanf($urlPath, '/album/%d/video/%d/%s');
				break;

			// Channel
			case 'channels':
				list ($channelId, $vimeoId, $unlistedHash) = sscanf($urlPath, '/channels/%d/%d/%s');
				break;

			// Group
			case 'groups':
				list ($groupId, $vimeoId, $unlistedHash) = sscanf($urlPath, '/groups/%d/videos/%d/%s');
				break;

			// On Demand video
			case 'ondemand':
				list ($ondemandId, $vimeoId, $unlistedHash) = sscanf($urlPath, '/ondemand/%d/%d/%s');
				break;

			// A regular video
			default:
				list ($vimeoId, $unlistedHash) = sscanf($urlPath, '/%d/%s');
				break;
		}

		return [
			'vimeoId' => $vimeoId,
			'unlistedHash' => $unlistedHash,
		];
	}
}
